server/src/main/java/io/deephaven/server/table/ops/WhereInGrpcImpl.java

```diff
@@ -0,0 +1,64 @@
+/**
+ * Copyright (c) 2016-2022 Deephaven Data Labs and Patent Pending
+ */
+package io.deephaven.server.table.ops;
+
+import io.deephaven.api.JoinMatch;
+import io.deephaven.base.verify.Assert;
+import io.deephaven.engine.table.Table;
+import io.deephaven.engine.updategraph.UpdateGraphProcessor;
+import io.deephaven.proto.backplane.grpc.BatchTableRequest;
+import io.deephaven.proto.backplane.grpc.TableReference;
+import io.deephaven.proto.backplane.grpc.WhereInRequest;
+import io.deephaven.server.session.SessionState;
+import io.deephaven.util.locks.AwareFunctionalLock;
+import io.grpc.StatusRuntimeException;
+
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import java.util.List;
+import java.util.Objects;
+
+@Singleton
+public class WhereInGrpcImpl extends GrpcTableOperation<WhereInRequest> {
+
+    private static List<TableReference> refs(WhereInRequest request) {
+        return List.of(request.getLeftId(), request.getRightId());
+    }
+
+    private final UpdateGraphProcessor updateGraphProcessor;
+
+    @Inject
+    public WhereInGrpcImpl(final UpdateGraphProcessor updateGraphProcessor) {
+        super(BatchTableRequest.Operation::getWhereIn, WhereInRequest::getResultId, WhereInGrpcImpl::refs);
+        this.updateGraphProcessor = Objects.requireNonNull(updateGraphProcessor);
+    }
+
+    @Override
+    public final void validateRequest(WhereInRequest request) throws StatusRuntimeException {
+        GrpcErrorHelper.checkHasField(request, WhereInRequest.LEFT_ID_FIELD_NUMBER);
+        GrpcErrorHelper.checkHasField(request, WhereInRequest.RIGHT_ID_FIELD_NUMBER);
+        GrpcErrorHelper.checkRepeatedFieldNonEmpty(request, WhereInRequest.COLUMNS_TO_MATCH_FIELD_NUMBER);
+    }
+
+    @Override
+    public final Table create(final WhereInRequest request, final List<SessionState.ExportObject<Table>> sourceTables) {
+        Assert.eq(sourceTables.size(), "sourceTables.size()", 2);
+        Assert.gtZero(request.getColumnsToMatchCount(), "request.getColumnsToMatchCount()");
+        final Table left = sourceTables.get(0).get();
+        final Table right = sourceTables.get(1).get();
+        final List<JoinMatch> columnsToMatch = JoinMatch.from(request.getColumnsToMatchList());
+        final AwareFunctionalLock lock =
+                left.isRefreshing() || right.isRefreshing() ? updateGraphProcessor.sharedLock() : null;
+        if (lock != null) {
```

Comment:
why explicit lock+unlock over computeLocked? at least as a reader the computeLocked is more concise, more obviously correct ("any chance that something could throw on the way out of lock() that would leave it locked, should we just do the lock inside the try/catch").

Example
io.deephaven.server.table.ops.JoinTablesGrpcImpl#create

```
        final Table result;
        if (!lhs.isRefreshing() && !rhs.isRefreshing()) {
            result = realTableOperation.apply(lhs, rhs, columnsToMatch, columnsToAdd, request);
        } else {
            result = updateGraphProcessor.sharedLock().computeLocked(
                    () -> realTableOperation.apply(lhs, rhs, columnsToMatch, columnsToAdd, request));
        }
        return result;
```
(could just as easily return from inside the if/else)

Reply:
I'm not a fan of lambda abstractions for singular operations (I see where the abstraction comes in handy for join where we have a lot of them), and the branching that results around the application. That said, the version I has has branching logic around the lock... something I personally prefer, but I think we can do better...

@rcaudy @niloc132 , would you be opposed to adding a new method to FunctionalLock:

```java
public interface FunctionalLock extends Lock {
    ...
    default SafeCloseable closeableLock() {
        final SafeCloseableLock closeable = new SafeCloseableLock(this);
        lock();
        return closeable;
    }
```

This would allow the use of a try resource, which I think is much more ergonomic:

```java
    @Override
    public final Table create(final WhereInRequest request, final List<SessionState.ExportObject<Table>> sourceTables) {
        ...
        try (final SafeCloseable _lock = lock(left, right)) {
            return request.getInverted() ? left.whereNotIn(right, columnsToMatch) : left.whereIn(right, columnsToMatch);
        }
    }

    private SafeCloseable lock(Table left, Table right) {
        return left.isRefreshing() || right.isRefreshing() ? updateGraphProcessor.sharedLock().closeableLock() : null;
    }
```

Reply:
I don't hate it. Not a bad option, for cases where lambdas are annoying.